### PR TITLE
HPCC-23183 Thread local hook fix when on main thread

### PR DESCRIPTION
**Pull request by jakesmith**
Track hooks installed on main thread separately, in a non
thread locl vector.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review.

**Comment by jakesmith** [comment]
@richardkchapman - modified so that callThreadTerminationHooks handles the condition based on main thread id.

File: system/jlib/jthread.cpp [hunk 1]
```diff
@@ -41,26 +41,39 @@
...
+static std::vector<ThreadTermFunc> mainThreadTermHooks;
+
+struct MainThreadIdHelper
+{
+    ThreadId tid;
+    MainThreadIdHelper()
+    {
+        tid = GetCurrentThreadId();
+    }
+} mainThreadIdHelper;
```
Comment (richardkchapman) [inline system/jlib/jthread.cpp]: should be static?

**Comment by richardkchapman** [comment]
One minor comment. Also would be worth adding a comment about why this is necessary (i.e. that thread-local objects are destroyed before functions called via atexit )

@ghalliday can you sanity check too

File: system/jlib/jthread.cpp (hunk 1 above)
Comment (jakesmith) [reply system/jlib/jthread.cpp]: agree, have changed.

**Comment by jakesmith** [comment]
made mainThreadIdHelper static and added a comment.

@ghalliday - please review.

File: system/jlib/jthread.cpp [hunk 2]
```diff
@@ -41,26 +41,45 @@
 
 //static __thread ThreadTermFunc threadTerminationHook;
```
Comment (richardkchapman) [inline system/jlib/jthread.cpp]: This line should have been removed by my previous PR in tis area..

**Pull request merged**